### ISSUE TITLE
TypeDecorator process_result_value behavior difference between 1.3 and 1.4 (or 2.0)

### ISSUE THREAD

**Issue by emscbQ** [sqlalchemy/sqlalchemy#11211]
### Describe the bug

I'm using custom `TypeDecorator` for converting json string to list of entity.

When loading the same entity, the `process_result_value` suddenly runs multiple times.
In version 1.3, the function was executed only once even if the same entity was loaded, but in 1.4, it is executed every time.

### SQLAlchemy Version in Use

1.3.17, 1.4.52, 2.0.29

### DBAPI (i.e. the database driver)

pymysql (sqlite in test)

### Database Vendor and Major Version

MySQL 8.0.28

### Python Version

3.11.6

### Operating system

Linux

### To Reproduce

```python
import json
import time
from dataclasses import asdict, dataclass
from typing import List

import jsons
import sqlalchemy
import sqlalchemy as sa
import sqlalchemy.orm as orm
from sqlalchemy import sql
from sqlalchemy.types import TEXT, TypeDecorator


class DataclassListJsonConverter(TypeDecorator):
    _custom_type = None
    _remove_protected: bool

    def __init__(self, custom_type):
        super().__init__()
        self._custom_type = custom_type

    @property
    def python_type(self):
        return self._custom_type

    def process_literal_param(self, value, dialect):
        pass

    impl = TEXT

    def process_bind_param(self, value, dialect):
        if value is None:
            return None
        elif not isinstance(value, List):
            raise Exception('only list allowed')

        value_dict_list = [asdict(i) for i in value]
        json_str = json.dumps(value_dict_list, ensure_ascii=False)
        return json_str

    def process_result_value(self, value, dialect):
        if value is None:
            return None

        start = time.time()
        json_dict_list = json.loads(value)
        obj_list = [jsons.load(i, self.python_type) for i in json_dict_list]
        end = time.time()
        if (elapsed := end - start) > 0.01:
            print(f"process_result_value {elapsed:.2f}s")
        return obj_list


@dataclass(frozen=True)
class Item:
    code: str


class Pool:
    id: str
    pooled_items: List[Item]

    def __init__(self, id: str, pooled_items: List[Item]):
        self.id = id
        self.pooled_items = pooled_items


def configure_pool(metadata: sa.MetaData):
    pool = sa.Table(
        'pool', metadata,
        sa.Column('id', sa.String(100), primary_key=True),
        sa.Column('pooled_items', DataclassListJsonConverter(Item)),
        sa.Column('created_at', sa.DateTime(), server_default=sql.func.now()),
        sa.Column('updated_at', sa.DateTime(), onupdate=sql.func.now()),
    )

    # 1.3
    orm.mapper(Pool, pool)

    # 1.4, 2.0
    mapper_registry = orm.registry()
    mapper_registry.map_imperatively(Pool, pool)


if __name__ == "__main__":
    engine = sqlalchemy.create_engine(
        'sqlite:///:memory:', echo=True
    )

    metadata = sqlalchemy.MetaData()
    configure_pool(metadata)

    session = sqlalchemy.orm.sessionmaker(bind=engine)()

    session.execute("CREATE TABLE pool ( id varchar, pooled_items varchar, created_at date, updated_at date )")
    session.add(Pool("1", [Item(f"{i}") for i in range(3000)]))
    session.commit()
    session.expunge_all()

    pools = session.query(Pool).all()
    pools = session.query(Pool).all()
```

### Error

Printed
process_result_value 0.22s
once in 1.3
double in 1.4, 2.0

**Comment by zzzeek** [comment]
hey there -

this is extensively documented in the [1.4 migration guide](https://docs.sqlalchemy.org/en/20/changelog/migration_14.html#proxying-behavior-goes-away-was-also-unnecessary-in-modern-usage) as part of the overall modernization of Core / ORM that took place in 1.4.

**Comment by emscbQ** [comment]
Thanks. However, by continuously calling process_result_value, the process progress time is increased.
Is there any way to speed up the process in my code?

**Comment by CaselIT** [comment]
Is there a reason why you keep quering the db? Trying to avoid that would be the best option

**Comment by zzzeek** [comment]
> Thanks. However, by continuously calling process_result_value, the process progress time is increased. Is there any way to speed up the process in my code?

I would not create dataclass instances inside of a TypeDecorator as this is an expensive process.   I would put a python descriptor on your model classes, possibly a memoized one, that generates the value from plain text only when called upon.

beyond that, as mentioned, changing query behavior is the other way, use `deferred()`, or query for just the individual columns you need and/or use `load_only()`.

**Comment by emscbQ** [comment]
Thanks! I roughly understood `deferred()` as lazy loading.

By using `deferred()`, the field is actually converted at the time of access.
When retrieving `Pool` the second time, the same pool object is returned.
But the conversion(call `process_result_value`) is not performed at access to field.
I wonder why it doesn't convert again.
